Gems/Atom/RPI/Assets/ShaderLib/Atom/RPI/TangentSpace.azsli

```diff
@@ -26,17 +26,22 @@ float3 WorldSpaceToTangent(float3 vectorWS, float3 normalWS, float3 tangentWS, f
     return float3(a,b,c);
 }
 
-//! Utility function for vertex shaders to transform vertex tangent, bitangent, and normal vectors into world space.
-void ConstructTBN(float3 vertexNormal, float4 vertexTangent, float3 vertexBitangent, float4x4 localToWorld, float3x3 localToWorldInverseTranspose, out float3 normalWS, out float3 tangentWS, out float3 bitangentWS)
+//! Utility function for *pixel* shaders to transform vertex tangent, bitangent, and normal vectors into world space.
+void ConstructTBN(float3 vertexNormal, float3 vertexTangent, float4x4 localToWorld, float3x3 localToWorldInverseTranspose, out float3 normalWS, out float3 tangentWS, out float3 bitangentWS)
 {
+    // No guarantee that normal and tangent will preserve unit length after interpolation in pixel shaders.
+    // Instead of doing normalization in each shader, do it here.
+    vertexNormal = normalize(vertexNormal);
+    vertexTangent = normalize(vertexTangent);
+    float3 vertexBitangent = cross(vertexTangent, vertexNormal);
```

Comment:
IIRC, it's possible for the geometry to use either left-handed or right-handed coordinates for the tangent space. This didn't matter before because the bitangent from the vertex input would be used for calculations. But since we are ignoring that now and reconstructing the bitangent instead, we need to consider the handedness. This should already be available in the tangent's w coordinate (that's why tangent is a float4 in VsInput).

Reply:
Thanks for pointing that out. I'll add a multiplier to the bitangent.